### ISSUE TITLE
Supercharger waypoint added manually: Can't adjust minimum stop time, SoC

### ISSUE THREAD

**Issue by EVGrokker** [aesculus/EVTO-App-Feedback#213]
Scenario:

1. Create a single-segment trip, Seattle to Moclips WA
2. Add Waypoint: Tesla Supercharger, Aberdeen WA
3. Route recalculates.
4. Touch Aberdeen supercharger pin to open Edit Waypoint dialog.

Minimum stop time is grayed out. Minimum Arrival SoC is grayed out.

Comments:
1. Minimum stop time should be enabled. Maybe user wants to treat this as a rest stop.
2. For waypoint stops at chargers, I think that the SoC slider should be 'Minimum Departure SoC'. 

Here's my thinking. 

If I decide to stop at a charger en route, my intent is to add charge, perhaps in anticipation of using the extra SoC for destination driving. Yes, there is another way to do this by reserving mileage in Edit Segment Details, but it's a reasonable user objective and a reasonable alternative approach. 

Another scenario would be unexpected consumption (detour, accident, sudden weather) en route, so user wants to add some SoC for insurance.

**Comment by aesculus** [comment]
So there are two things going on here:
- Not being able to get to disabled features
- Are the features correct and useful

I suggest we tackle the first one first, and then debate the second.

There are two reasons that these features may be disabled.
1. They are Pro features and the user has not paid for them. They may debate that they should be free because they can get it for free somewhere else. - maybe.  My choice I guess
2. They are disabled because of a design choice I made with auto chargers.

The later is what I think you are refering to because you have the Pro feature. I have purposefully disabled this on auto chargers because they are fluid. I place them and take them away based on conditions. There is no certainty that when you optimize you will get an auto charger again. It depends on the circumstances.

So if you changed the waypoint (Arrival SOC or Minimum Stop Time) and optimized again, all of your work could be thrown away. I thought that would frustrate people.

But if you want to do that, you just need to Require (now the pushpin) the charger (or it's downstream friend) and your settings and the charger will stay until you delete it.  And the route obviously will stay that way too.

So here are some options:
- Don't disable the feature and let people adjust to their heart's content knowing (probably not because they will never read or think of these contexts) the charger might disappear - and reappear later without their settings
- Don't disable them but if the user does adjust them, auto require the charger. They will have to remove the charger to really make the optimization automatic again. I suspect they won't understand that either.
- Cache the auto chargers and their state when they are no longer required just in case they show up again.  However the context might be different. Maybe the optimizer wants to charge for a longer or shorter time. Have not thought about how to deal with that. Interestingly today I do cache them while optimizing, but just during that process. I turn them off, optimize, and any that are still required get turned back on again with the new SOC requirements etc. The others are discarded.  I could just never discard them until you delete the trip. A possibility.

**Comment by EVGrokker** [comment]
With respect to the disabled Stop Time and SoC (sorry, should have created two issues): 
1. Disabled in standard version. I have no issue with this at all, as long as we call it out in the local help. That's the first place the user should look if they're puzzled. So let's consider that resolved, with a pending help text revision.

They are disabled because of a design choice I made with auto chargers.
The later is what I think you are refering to because you have the Pro feature. I have purposefully disabled this on auto chargers because they are fluid. I place them and take them away based on conditions. There is no certainty that when you optimize you will get an auto charger again. It depends on the circumstances.

>The later is what I think you are refering to because you have the Pro feature. I have purposefully disabled this on auto chargers because they are fluid. I place them and take them away based on conditions. There is no certainty that when you optimize you will get an auto charger again. It depends on the circumstances.
> So if you changed the waypoint (Arrival SOC or Minimum Stop Time) and optimized again, all of your work could be thrown away. I thought that would frustrate people.
> But if you want to do that, you just need to Require (now the pushpin) the charger (or it's downstream friend) and your settings and the charger will stay until you delete it. And the route obviously will stay that way too.

Actually, that's not the scenario I described. I had manually added the charger as a waypoint stop. So we're talking about either a manually added charging stop or an automatic charging stop that the user converted to a required, manual stop. My comments only apply to non-auto charging stops.

The bottom line for me is this: when the SoC slider is adjustable on a pinned (non-auto) charging stop, the wording should be 'Minimum Departure SoC', because that's what the user would be thinking in terms of - a booster charge.

**Comment by EVGrokker** [comment]
And to clarify my thoughts about 'should the slider be enabled for auto-added charging waypoints'… no.
As you say, far too much potential confusion. If they want to pin the stop, they can tweak to their heart's content.

**Comment by aesculus** [comment]
OK. Now we are moving to question 2, your real question.

So we have a manual waypoint. I see a few reasons to up the charge from what the app wants to do:
- The user wants more padding to make sure they get to the next stop. Or in your scenario they want some extra reserve for weather, local travel etc.  This is covered already by bumping up the Min Arrival SOC on the downstream charger. In fact I think my logic is better because I actually make them do what they want, increase the SOC buffer where it counts, at the next stop. I will admit it's seems confusing because you are going downstream to force upstream behavior. It's physics! And the other problem is that is what they are used to in the other planners. 
- The user wants to continue to charge for longer than required just because - maybe they are having a meal and the car is plugged in, so why not keep charging. 

I buy this scenario. The app does not accommodate it today. But what if I keep charging for the minimum stop time? ie if you are going to wait there, why not keep it charging for that amount of time? 

So by jacking up the minimum stop time at a charging stop, you will charge for as long as the minimum time or 100%.

> For waypoint stops at chargers, I think that the SoC slider should be 'Minimum Departure SoC'.

This would be a major change in the logic and also be confusing if I did not change this for all waypoints. Today the app looks at what you need upon arrival and does its best to make that happen. When you force a state like this at the start of a segment, you are playing games hoping that is enough to get you to the next stop. You then have to bounce back and forth to see if you got what you wanted.

Maybe I am just being too pedantic. It would not be the first time.

**Comment by EVGrokker** [comment]
I admit that I probably haven't thought through all of the possible ramifications here. I'm willing to spend some time thinking about this and compose a more thoughtful response.

To be clear, I do understand that user can specify reserving charge at the destination, but 1) there's nothing wrong with having two different ways to do that, depending on where you are in your planning, and 2) what about the scenario where the user adds two charging stops manually, and wants to pad the first stop? Destination reserve won't help there.

My gut is that non-charging waypoints should say 'Minimum Arrival SoC' (because they won't be recharging), and manual charging waypoints should say 'Minimum Departure SoC' (because that's under user control). But let me chew on this a bit and come back with a more polished argument.

**Comment by aesculus** [comment]
Sounds like a plan. I understand the anquist here for the first scenario. The user is wanting to do what they would do in the real world, charge longer because they are afraid or what might be ahead of them.

But the whole point of the app is to remove that anquist by having the app do the heaving lifting for them. I am supposed to look out for all of these conditions for them. In the end of the day for the padding scenario you really want the next charger (no matter how it is defined - manual or automatic) to get the padding, not the place you are charging at. The real need is that you want to "end up" safer, not start safer, IMHO.

**Comment by EVGrokker** [comment]
The more polished version:

Let's divide users into two groups - those who want the simplest, quickest projection of how to get from point A to point B using EVTO. They're using the app because it allows them to preview something similar to what the car would tell them. (Of course the car only does one waypoint at time.) These users will use the 'Auto' button when planning their new trip. They're not going to fuss with tweaking stop times, adjusting SoC buffers, etc. Let's call these the **auto users**.

If the auto user has adjusted settings to indicate that they don't like to drop below 20% SoC, or prefer having 50 miles of destination charge available upon arrival at their destination, auto users' trips will be planned with those defaults. But that's as deep as they're going to go. They're perfectly content with EVTO's predictions. Even if they plan to make a few stops along the way, the prediction is close enough to suit their purposes, and they're going to use the in-car navigation once they're underway. They just want some up-front hand-holding to reduce range anxiety and angst. These folks are already covered in the app as it exists today.

The second group we'll call the **manual users**. They're micromanaging their trip, thinking ahead about how long to allow at each waypoint. They like planning an accurate itinerary for their trips. These are the folks who want open the hood to tweak EVTO's calculations and get more trip planning features out of EVTO.

So let's plan a two-waypoint trip from a manual user's perspective. We have an origin, two waypoints, and a destination. Let's say the first waypoint will be a Starbucks/restroom stop (no charging), the second waypoint is an optional charging stop (not strictly required). Our Destination has a destination charger, but that doesn't influence the first segment's calculations. We want some buffer upon arrival, which we'll finetune as we see the consumption and predicted SoC graphs.

Origin will be Seattle WA, starting with a default full charge from my home charger. I input the Sand Dollar Inn & Cottages in Moclips WA with a NEMA 14-50 as my destination charger. At this point, I know that I can get there without recharging, but I want to stretch my legs and check out the Aberdeen supercharger and add some juice while I'm walking around.

1. New Trip | Seattle WA - Sand Dollar Inn & Cottages, Moclips WA. Destination charger: NEMA 14-50. Manual charger selection. Calculate. Now we have a simple Origin - Destination one-segment trip.
2. Add the Starbucks / restroom stop. Let's choose Tacoma for that. Starbucks, Pacific Avenue, Tacoma. I'm getting out of the car, let's allow 30 minutes for this stop. Touch the edit button, adjust Minimum Stop Time to 30 mins. 'Minimum Arrival SoC' is not adjusted.
3. Next stop, Aberdeen WA supercharger. Add Waypoint | Tesla Superchargers | Aberdeen WA. Because this is a charging stop, this is where the SoC should be relabeled 'Minimum Departure SoC'. I propose that if the user does not select a charger type, that label would remain 'Minimum Arrival SoC', as it does with the Starbucks stop. I go ahead and add the charging waypoint, and I notice that the arrival SoC and departure SoC as displayed in the Waypoint Inspector are the same values, because EVTO knows that I don't need to charge here. But I want to.

I understand that your position is that in order to induce EVTO to let me charge here, I should go to the next stop and indirectly influence the SoC required at Aberdeen by tweaking the Destination Reserve or tweaking the Minimal Arrival SoC for the Destination. But the user is not going to naturally think of those solutions. If you want to reserve energy for arrival, it makes sense to declare that in Edit Trip Details. But if I want to charge at a voluntary charging stop, I should be able to declare at that waypoint how much I intend to charge to before traveling on.

So, for manual trip planning with one or more waypoints, I'd suggest the following:
- Non-charging waypoints label the SoC slider 'Minimal arrival SoC' (as they are now). This means "how much of a charge I want to have upon arriving here". Defaults to Settings | SoC Reserve (is that true?), but I can manually edit it if I intend to do some local driving here.
- Creating a waypoint and selecting a charger type changes the SoC slider label to 'Minimal departure SoC'. This means I intend to charge at this charger to the level indicated. The default value of the SoC slider upon opening this dialog would be the expected arrival SoC that you're already showing in the battery icon for this waypoint in the inspector. It makes no sense to allow the user to set it lower than that level, but they can set it higher.
- Destinations should be labeled 'Minimum Arrival SoC (%) (as they are now). If the user want to preserve energy for destination excursions, they can do so either at this slider, or in the Edit Trip Details dialog using Dest. Reserve (mi).

As far as the underlying algorithm, I can't see how this approach would be that much different than what you're already factoring in. If there's something more complicated that I'm not appreciating, help me understand that.

I see users on TMC describing this approach frequently: "Whenever I stop at a charging stop, I charge to the level recommended by the car, then add 15-20% before proceeding to my next destination." That's what this proposal lets them do in EVTO, but ahead of time in the trip planning stages.

With all that said, let me circle back to your earlier comment:

>But the whole point of the app is to remove that anquist by having the app do the heaving lifting for them. I am supposed to look out for all of these conditions for them. In the end of the day for the padding scenario you really want the next charger (no matter how it is defined - manual or automatic) to get the padding, not the place you are charging at. The real need is that you want to "end up" safer, not start safer, IMHO.

Agreed 100% -- _for auto users_. My goal here is to extend the utility of the app to the manual users who want to micromanage their trip itinerary.

**Comment by EVGrokker** [comment]
Here's the trip ID:
Trip: Sand Dollar Inn & Cottages, [0oqoqstrv9uqc6wq1qc6ps]

**Comment by aesculus** [comment]
OK. But what if I just have the app charge at Aberdeen for as long as you want to be there for up to 100%?

So if you set at Minimum Stop Time for 50 minutes then I will charge for that amount of time up to 100%.

So any place that is a charger, if you specify a minimum stop I will charge for at least as long as you are there for.

BTW I guess we could not get caught up the penalties for charging longer than you need to but that is another story. My experiences is that I was never actually able to get to 100% at a SC after 2 hours, even when I got to 99% in an hour.

**Comment by EVGrokker** [comment]
You are correct, both scenarios make sense for a optional charging stop.

The user could adjust the Minimum Stop Time represent the amount of time they were willing to allow for a charging stop. EVTO would predict the departure SoC, which would be visible in the Waypoint Inspector. In this case the user is thinking "I want to take on whatever charge I can add in 50 minutes".

Or, they could go for a specific departure SoC, say 80%. "I want to charge to 80% before leaving charger."

From a trip planning perspective, before you've left home, I think it makes more sense to use the Minimum Departure SoC slider. From an en-route perspective, on the day of the trip, it makes more sense to use the Minimum Stop Time slider to guesstimate how the 50 minutes will affect their SoC at the subsequent stop.

They both make sense.

**Comment by aesculus** [comment]
SOC Pseudo Algorithm
For each driving instruction and summed for the waypoint 
{
•	Energy required to move the car from last waypoint
•	Weather impact energy
•	Elevation energy
•	Climate control energy
}
Now add:
•	Energy reserved for local travel (destination)
•	Energy needed to get to next charger (O/B or return to last charger)
Now we have a total energy required at this waypoint. All of this information was derived from all the factors of drive time, drive speed, weather, road conditions, 
Translate this energy to a SOC %. Let’s call this nSOC for needed SOC. 
•	Add the min arrival SOC (from default or the waypoint edit) mSOC
Now we know the total energy in battery SOC for this waypoint. Let’s call the tSOC = nSOC + mSOC. 
This is what you see in the Charge to SOC.
To determine how long we actually need to charge we have to know how much we had at the prior waypoint left over. This is either the origin SOC or the remainder (Min Arrival SOC from the prior waypoint). Lets call this aSOC.
So the total needed SOC change = tSOC – aSOC, but for level III chargers you need to know where you started from because it’s not linear. From here there are algorithms that determine how long it’s going to take to move from the aSOC to the nSOC state. 
Now you have the actual charge time. There is also 5 minutes added for setup/breakdown so the total time is the charge time plus setup time.

**Comment by aesculus** [comment]
BTW this I think is the same as this request I put in the other day to start this debate.

**Comment by EVGrokker** [comment]
This issue contains the 'Minimum Departure SoC' discussion. 

The following comment came from a user, requesting this ability:
>I currently have a short trip from Albert Lee to downtown Sioux Falls, and when I display the Worthington supercharger dialog box it currently shows that ill arrive with 36% of charge and only need to charge for 15 minutes. However what I know that it does not know is that I really want to charge 100% at the Worthington supercharger to accommodate what I plan to do the next morning. It would be very useful if there was a control on that dialog box that allowed me to override the charge to amount so that I could, in this case, set it to 100% and that way appropriate times would be calculated.

**Comment by EVGrokker** [comment]
Addressed by #103
